domain/access/state/permission.go

```diff
@@ -195,74 +198,167 @@ func (st *PermissionState) UpsertPermission(ctx context.Context, args access.Upd
 // for the given user on the given target.
 func (st *PermissionState) ReadUserAccessLevelForTarget(ctx context.Context, subject string, target corepermission.ID) (corepermission.Access, error) {
-	userAccessType := corepermission.NoAccess
+	userAccess := corepermission.NoAccess
 	db, err := st.DB()
 	if err != nil {
-		return userAccessType, errors.Trace(err)
+		return userAccess, errors.Trace(err)
+	}
+
+	user := dbPermissionUser{
+		Name: subject,
+	}
+	perm := dbPermission{
+		GrantOn: target.Key,
+	}
+
+	readQuery := `
+SELECT  (u.external) AS (&dbPermissionUser.*),
+        (p.access_type, p.uuid) AS (&dbPermission.*)
+FROM    v_user_auth u
+        LEFT JOIN v_permission p ON u.uuid = p.grant_to AND p.grant_on = $dbPermission.grant_on
+WHERE   u.name = $dbPermissionUser.name
+AND     u.disabled = false
+AND     u.removed = false
+`
+
+	readStmt, err := st.Prepare(readQuery, user, perm)
+	if err != nil {
+		return userAccess, errors.Annotatef(err, "preparing select user access level for target statement")
 	}
 
+	var baseExternalPerms dbPermission
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		userAccessType, err = st.userAccessLevel(ctx, tx, subject, target)
+		err = tx.Query(ctx, readStmt, user, perm).Get(&user, &perm)
+		if errors.Is(err, sql.ErrNoRows) {
+			return fmt.Errorf("%w for %q on %q", accesserrors.AccessNotFound, subject, target.Key)
+		} else if err != nil {
+			return errors.Annotatef(err, "reading user access level for target")
+		}
+
+		if user.External {
+			baseExternalPerms, err = st.baseExternalAccessForTarget(ctx, tx, target)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
 		return err
 	})
 	if err != nil {
-		return userAccessType, errors.Trace(err)
+		return userAccess, errors.Trace(domain.CoerceError(err))
+	}
+
+	userAccess = corepermission.Access(perm.AccessType)
+	if user.External && baseExternalAccessGreater(baseExternalPerms, userAccess) {
+		return corepermission.Access(baseExternalPerms.AccessType), nil
+	}
+	if perm.UUID != "" {
+		return userAccess, nil
+	}
+	return corepermission.NoAccess, fmt.Errorf("%w for %q on %q", accesserrors.AccessNotFound, subject, target.Key)
+}
+
+// ReadUserAccessLevelForTargetAddingMissingUser returns the user access level for
+// the given user on the given target. If the user is external and does not yet
+// exist, it is created. An accesserrors.AccessNotFound error is returned if no
+// access can be found for this user, and (only in the case of external users),
+// the everyone@external user.
+func (st *PermissionState) ReadUserAccessLevelForTargetAddingMissingUser(ctx context.Context, subject string, target corepermission.ID) (corepermission.Access, error) {
+	tag := names.NewUserTag(subject)
+	userAccess, err := st.ReadUserAccessLevelForTarget(ctx, subject, target)
+	if err == nil {
+		return userAccess, nil
+	} else if tag.IsLocal() || !errors.Is(err, accesserrors.AccessNotFound) {
```

Comment:
Add a comment to this line to help with readability.

Reply:
Added a comment and refactored boolean expression to make more sense.